### PR TITLE
CLOUDP-36569 Slicer methods

### PR DESCRIPTION
**Pull request by chunmingli**
https://jira.mongodb.org/browse/CLOUDP-36569

The HEAD backup agent slicer code is quite complicated and hard to reason about the concurrency semantics. I'm isolating the slicer here to be a simple interface. Later on will introduce a slice producer for the S3 interacting code to consume slices.

### CODE REVIEW

File: cps-pit/slicer.go [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+	Docs           chan bson.Raw
+	MinSliceSize   int
+	MaxSliceSize   int
+	Timeout        time.Duration
+	bufferDoc      bson.Raw
+	lastDocSent    OplogID
+	CompressorType string
+
+	errorChan    chan error
+	shutdownChan chan struct{}
```
Comment (chunmingli) [inline cps-pit/slicer.go]: I'm still looking into getting rid of this `shutdownChan`.

File: cps-pit/slice.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	}
+	return ret
+}
+
+func (slice *Slice) addDoc(doc bson.Raw) {
+	if doc == nil {
+		return
+	}
+	slice.NumDocs += 1
+	slice.UncompressedSize += len(doc)
```
Comment (lawctan) [inline cps-pit/slice.go]: len works here? not too familiar with go

File: cps-pit/slice.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+func (slice *Slice) done() error {
+	compressed, err := slice.placeholder.Get()
+	if err != nil {
+		return err
+	}
+
+	slice.Payload.Grow(len(compressed))
+	slice.Payload.Write(compressed)
+	return nil
```
Comment (lawctan) [inline cps-pit/slice.go]: how about adding a semantic such that after done() is called, calling addDoc() again will throw panic

File: cps-pit/slicer.go [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+	"sync"
+	"time"
+)
+
+type Slicer struct {
+	Docs           chan bson.Raw
+	MinSliceSize   int
+	MaxSliceSize   int
+	Timeout        time.Duration
+	bufferDoc      bson.Raw
```
Comment (lawctan) [inline cps-pit/slicer.go]: would be nice to have some brief documentation on what each field means

File: cps-pit/slicer.go [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+		MinSliceSize:   minSliceSize,
+		MaxSliceSize:   maxSliceSize,
+		CompressorType: compressor,
+		Timeout:        timeout,
+		lastDocSent:    startingPoint, // todo create a new lastDocSent
+		Logger:         logger,
+	}
+}
+
+func (slicer *Slicer) addBufferDoc(slice *Slice) {
```
Comment (lawctan) [inline cps-pit/slicer.go]: perhaps better named `transferBufferDocTo` ?

File: cps-pit/slicer.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+
+	timer := time.NewTimer(slicer.Timeout)
+	defer timer.Stop()
+
+	var doc bson.Raw
+OUTER:
+	for slice.UncompressedSize < slicer.MinSliceSize {
+		channelOpen := true
+		select {
+		case doc, channelOpen = <-slicer.Docs:
```
Comment (lawctan) [inline cps-pit/slicer.go]: confusing syntax lol. is the `=` necessary here?

File: cps-pit/slicer.go [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+	select {
+	case slicer.Docs <- doc:
+		return true
+	case <-slicer.shutdownChan:
+		return false
+	}
+}
+
+func (slicer *Slicer) CreateSlice() (*Slice, error) {
```
Comment (lawctan) [inline cps-pit/slicer.go]: add doc to say that we create slice if one of the ff is true: exceeded MaxSliceSize or timer is up

File: cps-pit/slicer.go (hunk 6 above)
Comment (chunmingli) [reply cps-pit/slicer.go]: Assignment is necessary. This is receiving from a channel `x := <-someChan`, not sending to a channel `someChan <- x`.

File: cps-pit/slice.go (hunk 2 above)
Comment (chunmingli) [reply cps-pit/slice.go]: bson.Raw is defined as `type Raw []byte` and `len()` works with array/slice types.

File: cps-pit/slice.go (hunk 3 above)
Comment (chunmingli) [reply cps-pit/slice.go]: Hmm. Ok. I'll unset the `placeholder` (to nil) in this case as compressed bytes are written to the payload.

**Pull request merged**